consensus/polybft/consensus_runtime.go

```diff
@@ -325,7 +342,11 @@ func (c *consensusRuntime) populateFsmIfBridgeEnabled(
 func (c *consensusRuntime) FSM() error {
 	// figure out the parent. At this point this peer has done its best to sync up
 	// to the head of their remote peers.
-	parent, epoch := c.getLastBuiltBlockAndEpoch()
+	c.lock.RLock()
+	proposerSnapshot := c.proposerSnapshot.Copy()
+	parent, epoch := new(types.Header), new(epochMetadata)
```

Comment:
I am not sure about this, because now we are not consistent, since we have that `getEpoch` and `getLastBlockAndEpoch` functions which return copied data. 

What do you think, is it better to have a single function that returns all three fields copied, which is under Read lock, something like `getEpochFields() (*proposerSnapshot, *types.Header, *epochMetadata)`, so we can use it everywhere, since there are places in runtime that require latest block and epoch, some places where we require only block, some where we require all three, some where we require only epoch, or to just use the read lock and copy only what we need in the function that needs it.

Both have their advantages and disadvantages, but my point is we should just choose one approach and be consistent. My vote is for the first option since we don't hold that much data in those fields :thinking:

Reply:
We need to have something like 
```
type runtimeSyncData struct {
	// epoch is the metadata for the current epoch
	epoch *epochMetadata

	// lastBuiltBlock is the header of the last processed block
	lastBuiltBlock *types.Header

	// proposerSnapshot is the object which calculates new proposer
	proposerSnapshot *ProposerSnapshot
}
```

Reply:
Fixed on https://github.com/0xPolygon/polygon-edge/pull/1047 https://github.com/0xPolygon/polygon-edge/pull/1047/commits/751644154d6a0ff5f1d3cd048534f636c45f7cb9